### PR TITLE
Redo coercion site code

### PR DESCRIPTION
**Pull request by philberty**
This gets rid of the old visitor method and brings us much closer to the
Rustc rules which from the algo mentioned in the comment's do the checks
in a very specific order which we need to match.

### CODE REVIEW

File: gcc/rust/typecheck/rust-coercion.cc [hunk 1]
```diff
@@ -22,59 +22,275 @@ namespace Rust {
...
+      needs_reborrow = true;
+      expected_mutability = to_mutbl;
+
+      adjustments.push_back (
+	Adjustment (Adjustment::AdjustmentType::INDIRECTION, source_ref, ty_a));
+    }
+
+  // there is a bunch of code to ensure something is coerce able to a dyn trait
+  // we need to support but we need to support a few more lang items for that
+  // see:
```
Comment (dafaust) [inline gcc/rust/typecheck/rust-coercion.cc:249]: Might be good to stick a FIXME on this

**Review by dafaust — APPROVED** [review]
LGTM I like how you've split up the coercion functions. Very easy to follow.
Love to see all the deletions :)

File: gcc/rust/typecheck/rust-coercion.cc (hunk 1 above)
Comment (philberty) [reply gcc/rust/typecheck/rust-coercion.cc:250]: Fixed :D

**Comment by philberty** [comment]
Thanks for the review @dafaust I think with some tweaks we can do something similar for casts next which will be exciting :D.

**Pull request merged**